source/common/upstream/upstream_impl.cc

```diff
@@ -864,9 +864,12 @@ bool BaseDynamicClusterImpl::updateDynamicHostList(const HostVector& new_hosts,
 
     bool found = false;
     for (auto i = current_hosts.begin(); i != current_hosts.end();) {
-      // If we find a host matched based on address, we keep it. However we do change weight inline
-      // so do that here.
-      if (*(*i)->address() == *host->address()) {
+      const bool address_matched = *(*i)->address() == *host->address();
+      const bool health_check_changed =
+          health_checker_ != nullptr && *(*i)->healthCheckAddress() != *host->healthCheckAddress();
+      // If we find a host matched based on address and the health check address is not changed, we
+      // keep it. However we do change weight inline so do that here.
+      if (address_matched && !health_check_changed) {
```

Comment:
It's not your fault, but this entire method is completely unreadadble to me; it's hard to keep track of what `current_hosts` represent, which hosts are to be removed, under which conditions we're going to do a rebuild, inplace modified or remove/add. Do you think there's a way to explain to the layman how this works? Each time I reread this method, by head hurts more and more (and I'm partly to blame for sure).

Reply:
Yes I agree. This has been an iterative process over time and I agree it's now not very readable and it hurts my head also. I think it would be worth it to have a lot more comments in here, possibly variable name changes, and maybe even functions broken out. @dio since you are in here now do you mind taking a look?

Reply:
> since you are in here now do you mind taking a look?

Sure, love to have a look 🙂. Will spend some time dive into this.

Reply:
It seems there is a quite intensive refactoring in here: https://github.com/envoyproxy/envoy/pull/3959